File: src/leibniz-simd.cs
using System.Runtime.Intrinsics;

var rounds = uint.Parse(File.ReadAllText("rounds.txt"));
var unroll = (uint)Vector512<double>.Count;

var den   = Vector512<double>.Zero;
var inc   = Vector512.Create((double)unroll);
var two   = Vector512.Create(2.0);
var mone  = Vector512.Create(-1.0);
var xvec  = Vector512.Create(-1.0, 1.0, -1.0, 1.0, -1.0, 1.0, -1.0, 1.0);
var ivec  = Vector512.Create(2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0);
var pivec = Vector512<double>.Zero;

rounds += 2;
var vend = rounds - (rounds % unroll);

for (var i = 2u; i < vend; i += unroll) {
    den   = (two * ivec) + mone;
    ivec  += inc;
    pivec += xvec / den;
}

var x  = 1.0D;
var pi = 1.0D + Vector512.Sum(pivec);

for (var i = vend; i < rounds; ++i) {
    x = -x;
    pi += x / (2 * i - 1);
}

pi *= 4;
Console.WriteLine(pi);

File: src/leibniz-simd.swift
import Glibc

func readRounds() -> UInt64 {
    var rounds: UInt64 = 0
    guard let file = fopen("rounds.txt", "r") else {
        fatalError("Cannot open rounds.txt")
    }
    defer { fclose(file) }
    if fscanf(file, "%llu", &rounds) != 1 {
        fatalError("Cannot parse rounds.txt")
    }
    return rounds
}

var rounds = readRounds()

var pi: Double = 1.0

let roundsRemainder = rounds % 4
rounds -= roundsRemainder

rounds += 2 // do this outside the loop

var i: UInt = 2
while i < rounds {
    let xVec = SIMD4<Double>(-1.0, 1.0, -1.0, 1.0)
    let iVec = SIMD4<Double>(Double(i) + 0, Double(i) + 1, Double(i) + 2, Double(i) + 3)
    pi += (xVec / (2.0 * iVec - 1.0)).sum()
    i += 4
}

for _ in 0..<roundsRemainder {
    let x = -1.0 + 2.0 * Double(i & 1)
    pi += x / Double(2 * i - 1)
    i += 1
}

pi *= 4.0
printf("%.16f\n", pi)
